include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+    std::optional<std::string> signal_name;
+
+    static data_sink_query
+    with_signal_name(std::string_view name) {
+        return { {}, std::string{ name } };
+    }
+
+    static data_sink_query
+    with_sink_name(std::string_view name) {
+        return { std::string{ name }, {} };
+    }
+};
+
+class data_sink_registry {
+    std::mutex            _mutex;
+    std::vector<std::any> _sinks;
+
+public:
+    // TODO this shouldn't be a singleton but associated with the flow graph (?)
+    static data_sink_registry &
+    instance() {
+        static data_sink_registry s_instance;
+        return s_instance;
+    }
+
+    template<typename T>
+    void
+    register_sink(data_sink<T> *sink) {
+        std::lock_guard lg{ _mutex };
+        _sinks.push_back(sink);
+    }
+
+    template<typename T>
+    void
+    unregister_sink(data_sink<T> *sink) {
+        std::lock_guard lg{ _mutex };
+        std::erase_if(_sinks, [sink](const std::any &v) {
+            try {
+                return std::any_cast<data_sink<T> *>(v) == sink;
+            } catch (...) {
+                return false;
+            }
+        });
+    }
+
+    template<typename T>
+    std::shared_ptr<typename data_sink<T>::poller>
+    get_streaming_poller(const data_sink_query &query, blocking_mode block = blocking_mode::NonBlocking) {
+        std::lock_guard lg{ _mutex };
+        auto            sink = find_sink<T>(query);
+        return sink ? sink->get_streaming_poller(block) : nullptr;
+    }
+
+    template<typename T, TriggerPredicate P>
+    std::shared_ptr<typename data_sink<T>::dataset_poller>
+    get_trigger_poller(const data_sink_query &query, P p, std::size_t pre_samples, std::size_t post_samples, blocking_mode block = blocking_mode::NonBlocking) {
+        std::lock_guard lg{ _mutex };
+        auto            sink = find_sink<T>(query);
+        return sink ? sink->get_trigger_poller(std::forward<P>(p), pre_samples, post_samples, block) : nullptr;
+    }
+
+    template<typename T, TriggerObserverFactory F>
+    std::shared_ptr<typename data_sink<T>::dataset_poller>
+    get_multiplexed_poller(const data_sink_query &query, F triggerObserverFactory, std::size_t maximum_window_size, blocking_mode block = blocking_mode::NonBlocking) {
+        std::lock_guard lg{ _mutex };
+        auto            sink = find_sink<T>(query);
+        return sink ? sink->get_multiplexed_poller(std::forward<F>(triggerObserverFactory), maximum_window_size, block) : nullptr;
+    }
+
+    template<typename T, TriggerPredicate P>
+    std::shared_ptr<typename data_sink<T>::dataset_poller>
+    get_snapshot_poller(const data_sink_query &query, P p, std::chrono::nanoseconds delay, blocking_mode block = blocking_mode::NonBlocking) {
+        std::lock_guard lg{ _mutex };
+        auto            sink = find_sink<T>(query);
+        return sink ? sink->get_snapshot_poller(std::forward<P>(p), delay, block) : nullptr;
+    }
+
+    template<typename T, typename Callback>
```

Comment:
`Callback` should be a constraint -> here, a concept would be helpful for users to communicate which signature is needed.

Same for the occurrences below.

Reply:
Done